### ISSUE TITLE
Pointer (even if it's const) can be reassigned to non-pointer

### ISSUE THREAD

**Issue by dbandstra** [ziglang/zig#3649]
This nonsense code successfully compiles:
```zig
const Struct = struct { field: u32 };
pub fn main() void {
    var val = Struct { .field = 1 };
    const ptr = &val;
    ptr = Struct { .field = 2 };

    @import("std").debug.assert(ptr == &val);
}
```

It will correctly fail if you change the struct to a plain u32, or if you try to reassign ptr to a pre-existing variable instead of a literal.

A variation - this segfaults the compiler:
```zig
const Struct = struct { field: u32 };
pub fn main() void {
    // val is now const, ptr is now var
    const val = Struct { .field = 1 };
    var ptr = &val;
    ptr = Struct { .field = 2 };
}   
```

**Comment by momumi** [comment]
Another variant using slices:

```
const warn = @import("std").debug.warn;

pub fn main() anyerror!void {
    var odd_nums = [3]u8 {1,3,5};
    var even_nums = [3]u8 {2,4,6};
    const slice: []u8 = even_nums[0..];
    slice[0] = 0;
    slice = odd_nums[0..]; // should be invalid, slice is const
    slice[0] = 0;

    warn("odds : {x}\n", .{ odd_nums }); // prints  000305
    warn("evens: {x}\n", .{ even_nums }); // prints 000406
}
```

**Comment by Vexu** [comment]
Fixed by #4975.

```
./build/a.zig:267:18: error: type '*const Struct' does not support struct initialization syntax
    ptr = Struct { .field = 2 };
                 ^
./build/a.zig:267:21: error: cannot assign to constant
    slice = odd_nums[0..]; // should be invalid, slice is const
                    ^
./build/a.zig:264:20: error: cannot assign to constant
    ptr.* = Struct { .field = 2 }; // should be invalid, since ptr points to const data
                   ^
./build/a.zig:267:17: error: cannot assign to constant
    ptr.field = 2; // this fails to compile (correct behavior)
                ^
./build/a.zig:270:10: error: cannot assign to constant
    a = U{ .B = 1 }; // invalid, a is const
         ^
```
